parquet/src/arrow/async_reader/store.rs

```diff
@@ -45,29 +46,35 @@ use tokio::runtime::Handle;
 /// println!("Found Blob with {}B at {}", meta.size, meta.location);
 ///
 /// // Show Parquet metadata
-/// let reader = ParquetObjectReader::new(storage_container, meta);
+/// let reader = ParquetObjectReader::new(storage_container, meta.location, Some(meta.size));
 /// let builder = ParquetRecordBatchStreamBuilder::new(reader).await.unwrap();
 /// print_parquet_metadata(&mut stdout(), builder.metadata());
 /// # }
 /// ```
 #[derive(Clone, Debug)]
 pub struct ParquetObjectReader {
     store: Arc<dyn ObjectStore>,
-    meta: ObjectMeta,
+    path: Path,
+    file_size: Option<usize>,
     metadata_size_hint: Option<usize>,
     preload_column_index: bool,
     preload_offset_index: bool,
     runtime: Option<Handle>,
 }
 
 impl ParquetObjectReader {
-    /// Creates a new [`ParquetObjectReader`] for the provided [`ObjectStore`] and [`ObjectMeta`]
+    /// Creates a new [`ParquetObjectReader`] for the provided [`ObjectStore`] and [`Path`].
     ///
-    /// [`ObjectMeta`] can be obtained using [`ObjectStore::list`] or [`ObjectStore::head`]
-    pub fn new(store: Arc<dyn ObjectStore>, meta: ObjectMeta) -> Self {
+    /// The file size is optional, and if provided, it will ensure that only bounded range requests
+    /// are used. If file size is not provided, the reader will use suffix range requests to fetch
+    /// the metadata.
+    ///
+    /// The file size can be obtained using [`ObjectStore::list`] or [`ObjectStore::head`].
+    pub fn new(store: Arc<dyn ObjectStore>, path: Path, file_size: Option<usize>) -> Self {
```

Comment:
What do you think about not including the optional`file_size` in the constructor, and instead add a `with_file_size` method? I think that might be more consistent with the `metadata_size_hint` and the other options on this struct.

```suggestion
    pub fn new(store: Arc<dyn ObjectStore>, path: Path) -> Self {
```

And then you would make one of these things like

```rust
        let reader = ParquetObjectReader::new(store, meta.location)
            .with_size(Some(meta.size))
            .with_runtime(rt.handle().clone());
```

I think as long as we left  documentation that clearly stated that providing the size up front was an important optimization to avoid extra calls, this would be ok

Reply:
Closed in [`6b94524` (#7334)](https://github.com/apache/arrow-rs/pull/7334/commits/6b9452422da698db9cea669c539bbd392c69118b)